### PR TITLE
Add number of directories to FuseIOBench

### PR DESCRIPTION
**Pull request by ssz1997**
The number of files must be a multiple of number of directories, so each directory has same amount of files.
The number of files must be a multiple of number of threads, so each thread processes same amount of files.

@LuQQiu PTAL. Much cleaner than last time. Thanks!

### CODE REVIEW

File: stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java [hunk 1]
```diff
@@ -68,11 +70,16 @@ public static void main(String[] args) {
 
   @Override
   public void prepare() throws Exception {
+    if (mParameters.mNumFiles % mParameters.mNumDirs != 0) {
+      throw new IllegalArgumentException(String
+          .format("Number of files (%d) must be a multiple of the number of directories (%d)",
```
Comment (LuQQiu) [inline stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java:75]: I don't like the idea of forcing users to put two parameters that must be a multiplier. Every user will not be happy seeing that. 
What i recommend before is 
- we ask users to provide numDirs, numFilesPerDir, numDirs is better to be a multiplier of the thread number, but it's not mandary. 

Yes, we can simplify our developing logic by making users a little harder to use, but making them super hard to use should be avoided.

File: stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java [hunk 2]
```diff
@@ -68,11 +70,16 @@ public static void main(String[] args) {
...
+              mParameters.mNumFiles, mParameters.mNumDirs));
+    }
     for (Integer numOfThreads: mParameters.mThreads) {
-      if (numOfThreads > mParameters.mNumFiles) {
+      if (mParameters.mNumFiles % numOfThreads != 0) {
```
Comment (LuQQiu) [inline stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java:79]: In our case, better to not allowing thread array list. 
I took a look at Fuse logics, there are many operating system cache involved in reading, which means that when the application read a file for the first time, it goes to Fuse, but then the file may be cached in the operating system and no actual read happens at AlluxioFuse. 
We should just change mThreads to a single integer

**Review by LuQQiu — CHANGES_REQUESTED** [review]

File: stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java (hunk 1 above)
Comment (ssz1997) [reply stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java:75]: I see. Yeah this sounds more reasonable. Thanks!

**Comment by ssz1997** [comment]
@LuQQiu Each directory is only written by one thread. PTAL thanks!

File: stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java [hunk 3]
```diff
@@ -293,23 +288,25 @@ private void runInternal() throws Exception {
...
-          if (CommonUtils.getCurrentMs() > recordMs) {
-            if (ioBytes > 0) {
-              mThreadCountResult.incrementIOBytes(ioBytes);
-            } else {
-              // Done reading/writing one file
-              break;
+      for (int dirId = mThreadId; dirId < mParameters.mNumDirs; dirId += mParameters.mThreads) {
+        for (int fileId = 0; fileId < mParameters.mNumFilesPerDir; fileId++) {
+          mCurrentOffset = 0;
+          String filePath = mParameters.mLocalPath + "/" + dirId + "/" + fileId;
```
Comment (LuQQiu) [inline stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java:294]: Avoid string concatenation, which is low efficiency. Each + will create a new object.
use string.format here, in other cases, can use stringBuilder

File: stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java [hunk 4]
```diff
@@ -293,23 +288,25 @@ private void runInternal() throws Exception {
...
+            }
+            long ioBytes = applyOperation(filePath);
+
+            // Start recording after the warmup
+            if (CommonUtils.getCurrentMs() > recordMs) {
+              if (ioBytes > 0) {
+                mThreadCountResult.incrementIOBytes(ioBytes);
+              } else {
+                // Done reading/writing one file
+                break;
```
Comment (LuQQiu) [inline stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java:308]: if (CommonUtils.getCurrentMs() < recordMs && ioBytes <= 0 ) { we should also break! we also done reading/writing one file
Move the break logic after long ioBytes = applyOperation(filePath);

**Review by LuQQiu — CHANGES_REQUESTED** [review]

File: stress/common/src/main/java/alluxio/stress/fuse/FuseIOParameters.java [hunk 5]
```diff
@@ -48,9 +45,14 @@
...
-  @Parameter(names = {"--num-files"},
-      description = "The number of files for the test")
-  public int mNumFiles = 1000;
+  @Parameter(names = {"--num-files-per-dir"},
+      description = "The number of files per directory for the test")
+  public int mNumFilesPerDir = 1000;
+
+  @Parameter(names = {"--num-dirs"},
+      description = "The number of directories that the files will be evenly distributed into."
+          + "It is suggested to be a multiple of the number of threads.")
```
Comment (LuQQiu) [inline stress/common/src/main/java/alluxio/stress/fuse/FuseIOParameters.java:54]: The number of directories should be bigger or equal to the thread number, and it's suggested to be a multiple of the number of threads.

File: stress/common/src/main/java/alluxio/stress/fuse/FuseIOParameters.java [hunk 6]
```diff
@@ -48,9 +45,14 @@
           description = "The buffer size for IO operations. (1k, 16k, etc.)")
   public String mBufferSize = "64k";
 
-  @Parameter(names = {"--num-files"},
-      description = "The number of files for the test")
-  public int mNumFiles = 1000;
+  @Parameter(names = {"--num-files-per-dir"},
+      description = "The number of files per directory for the test")
```
Comment (LuQQiu) [inline stress/common/src/main/java/alluxio/stress/fuse/FuseIOParameters.java:49]: no need to add `for the test`

File: stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java [hunk 7]
```diff
@@ -68,12 +70,9 @@ public static void main(String[] args) {
...
   @Override
   public void prepare() throws Exception {
-    for (Integer numOfThreads: mParameters.mThreads) {
-      if (numOfThreads > mParameters.mNumFiles) {
-        throw new IllegalArgumentException(String
-            .format("Number of threads (%d) must be larger than number of files (%d)",
-                numOfThreads, mParameters.mNumFiles));
-      }
+    if (mParameters.mThreads > mParameters.mNumDirs) {
+      LOG.warn("Some threads are not being used. It is suggested to make the number of "
```
[the remark this answers is not in the document]
Comment (ssz1997) [reply stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java:74]: I'm going to throw an exception. Setting mThreads = mNumDirs without telling users explicitly would make them confused at the end why the mThreads was changed.

File: stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java [hunk 8]
```diff
@@ -291,22 +293,21 @@ private void runInternal() throws Exception {
...
+          String filePath = String.format("%s/%d/%d", mParameters.mLocalPath, dirId, fileId);
           while (!Thread.currentThread().isInterrupted()) {
             if (isRead && CommonUtils.getCurrentMs() > mContext.getEndMs()) {
               closeInStream();
               return;
             }
             long ioBytes = applyOperation(filePath);
 
+            // Done reading/writing one file
+            if (ioBytes < 0) {
```
Comment (LuQQiu) [inline stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java:305]: <= 0

**Review by LuQQiu — COMMENTED** [review]
Overall LGTM, careful about all the corner cases, thanks!

**Comment by ssz1997** [comment]
@LuQQiu Fixed the comments. PTAL thanks!

**Review by LuQQiu — APPROVED** [review]
LGTM, thanks for supporting writing millions of files through Fuse stress bench test

**Comment by ssz1997** [comment]
alluxio-bot, merge this please

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

**Pull request merged**